[ElimAvailExtern] Add an option to allow to convert global variables in a specified address space to local

shiltian opened the pull request:
Currently, the `EliminateAvailableExternallyPass` only converts certain
available externally functions to local if `avail-extern-to-local` is set or in
contextual profiling mode. For global variables, it only drops their
initializers.

This PR adds an option to allow the pass to convert global variables in a
specified address space to local. The motivation for this change is to correctly
support lowering of LDS variables (`__shared__` variables, in more generic
terminology) when ThinLTO is enabled for AMDGPU.

A `__shared__` variable is lowered to a hidden global variable in a particular
address space by the frontend, which is roughly same as a `static` local
variable. To properly lower it in the backend, the compiler needs to check all
its uses. Enabling ThinLTO currently breaks this when a function containing a
`__shared__` variable is imported from another module. Even though the global
variable is imported along with its associated function, and the function is
privatized by the `EliminateAvailableExternallyPass`, the global variable itself
is not.

It's safe to privatize such global variables, because they're _local_ to their
associated functions. If the function itself is privatized, its associated
global variables should also be privatized accordingly.

shiltian commented:
* **#144287** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/144287?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/144287?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

mtrofin commented on llvm/lib/Transforms/IPO/ElimAvailExtern.cpp:
```diff
@@ -35,8 +35,14 @@ static cl::opt<bool> ConvertToLocal(
+
 STATISTIC(NumRemovals, "Number of functions removed");
-STATISTIC(NumConversions, "Number of functions converted");
+STATISTIC(NumConversions, "Number of functions and globals converted");
```
would it help if we split this statistic in `NumFunctionsConverted` and `NumGlobalVariablesConverted`?

shiltian replied:
Sure, will do

shiltian commented on llvm/lib/Transforms/IPO/ElimAvailExtern.cpp:
```diff
@@ -88,9 +94,32 @@ static void convertToLocalCopy(Module &M, Function &F) {
 
+  // Convert global variables in specified address space before changing it to
+  // external linkage below.
+  if (ConvertGlobalVariableInAddrSpace.getNumOccurrences()) {
```
Sure, although that means we will have to have the if-statement in each iteration.

mtrofin replied:
I doubt it - likely some loop optimization would handle that; if it appears on some compile-time benchmark, we can revisit.

teresajohnson reviewed (commented):
> Currently, the EliminateAvailableExternallyPass only converts certain available externally functions to local.

This is not the default behavior. This pass mostly exists to drop definitions, not to convert to local. The conversion was added for contextual profiling. So the reference to this behavior in the summary is a bit odd. Do you *also* want to convert functions to locals? If so, should this be under the same option(s)?

shiltian commented:
> This is not the default behavior. This pass mostly exists to drop definitions, not to convert to local. The conversion was added for contextual profiling. So the reference to this behavior in the summary is a bit odd.

Thanks for the information. I'll correct the description.

>  Do you _also_ want to convert functions to locals?

Yes but that is controlled by the existing option `avail-extern-to-local`.

> If so, should this be under the same option(s)?

My understanding is `avail-extern-to-local` is for function and the new option is for global variable. Do you think it's a better idea to check both options for global variables?

teresajohnson commented:
> > This is not the default behavior. This pass mostly exists to drop definitions, not to convert to local. The conversion was added for contextual profiling. So the reference to this behavior in the summary is a bit odd.
> 
> Thanks for the information. I'll correct the description.
> 
> > Do you _also_ want to convert functions to locals?
> 
> Yes but that is controlled by the existing option `avail-extern-to-local`.

I guess there are 2 cases that we convert functions to locals, one is for contextual profiling (sets the Convert bool) and one is if the `avail-extern-to-local` flag is given.

> 
> > If so, should this be under the same option(s)?
> 
> My understanding is `avail-extern-to-local` is for function and the new option is for global variable. Do you think it's a better idea to check both options for global variables?

I think it is fine to keep them separate for now, please just update the description.

shiltian committed:
[ElimAvailExtern] Add an option to allow to convert global variables in a specified address space to local

Currently, the `EliminateAvailableExternallyPass` only converts certain
available externally functions to local. For global variables, it only drops
their initializers.

This PR adds an option to allow the pass to convert global variables in a
specified address space to local. The motivation for this change is to correctly
support lowering of LDS variables (`__shared__` variables, in more generic
terminology) when ThinLTO is enabled for AMDGPU.

A `__shared__` variable is lowered to a hidden global variable in a particular
address space by the frontend, which is roughly same as a `static` local
variable. To properly lower it in the backend, the compiler needs to check all
its uses. Enabling ThinLTO currently breaks this when a function containing a
`__shared__` variable is imported from another module. Even though the global
variable is imported along with its associated function, and the function is
privatized by the `EliminateAvailableExternallyPass`, the global variable itself
is not.

It's safe to privatize such global variables, because they're _local_ to their
associated functions. If the function itself is privatized, its associated
global variables should also be privatized accordingly.

mtrofin commented on llvm/lib/Transforms/IPO/ElimAvailExtern.cpp:
```diff
@@ -45,6 +52,10 @@ void deleteFunction(Function &F) {
 }
 
+static std::string getNewName(Module &M, const GlobalValue &GV) {
+  return GV.getName().str() + ".__uniq" + getUniqueModuleId(&M);
```
nit: I meant also pull into the refactoring the setting of the name and the setting of the linkage type - basically reusing `convertToLocalCopy`

shiltian replied:
I tried that route but not sure if that would be better. The new name is used at L85 as well. If we are gonna have a function that does both rename and set linkage, it also needs to return the new name for L85, which I find the semantics of it weird.

mtrofin reviewed (approved):
lgtm, some nits.

also note the github-actions warning about undef / poison.

teresajohnson commented:
I have a couple questions/comments about the description:

> A `__shared__` variable is lowered to a hidden global variable in a particular address space by the frontend, which is roughly same as a `static` local variable. To properly lower it in the backend, the compiler needs to check all its uses. Enabling ThinLTO currently breaks this when a function containing a `__shared__` variable is imported from another module. > Even though the global variable is imported along with its associated function, and the function is privatized by the `EliminateAvailableExternallyPass`, the global variable itself is not.

This should maybe be "if the function is privatized by ..." because that's only optional behavior. 

> It's safe to privatize such global variables, because they're _local_ to their associated functions. If the function itself is privatized, its associated global variables should also be privatized accordingly.

This is only safe *if* the function is also privatized, correct? Should the new option only take effect if the Convert bool is set (ensuring functions are also privatized)? And possibly assert if not?

shiltian commented:
> also note the github-actions warning about undef / poison.

Yeah, but I think we'd want to keep that `undef` since it is the initializer for the global variable.

shiltian commented:
> This is only safe if the function is also privatized, correct? Should the new option only take effect if the Convert bool is set (ensuring functions are also privatized)? And possibly assert if not?

At a high level (from a target-independent perspective), these are separate. We want to provide an option to privatize global variables in a specified address space, that's all. For the AMDGPU LDS variable scenario specifically, it should be the driver's responsibility to make sure the correct combination of options is used.
